rfcs/convergence/make-styles-no-functions.md

```diff
@@ -0,0 +1,94 @@
+# Problem
+
+Currently `makeStyles()` can define style rules in two different ways
+
+```ts
+makeStyles({
+  rootA: { color: 'red' },
+  rootB: theme => ({ color: theme.tokenB }),
+});
+```
+
+`theme` is typed and coupled to tokens from `@fluentui/react-theme`. This tokens shape is not extensible for customers.
+
+# Solution
+
+## Export tokens separately
+
+Tokens are just CSS variables, we communicate this fact to customers so that they understand clearly what they are using. This also removes need in `useTheme()` hook as tokens can be accessed directly.
+
+```tsx
+import { tokens } from '@fluentui/react-theme';
+
+function CustomComponent() {
+  return <div style={{ color: tokens.tokensA }} />;
+}
+```
+
+`tokens` is just a plain object:
+
+```ts
+import type { Theme } from '@fluentui/react-theme';
+
+const tokens: Theme = {
+  borderRadiusNone: 'var(--borderRadiusNone)',
+  borderRadiusSmall: 'var(--borderRadiusSmall)',
+  /* ... */
+};
+```
+
+## Remove functions
+
+Once `tokens` are available there is no more need for functional style rules in `makeStyles()`:
+
+```diff
+import { makeStyles } from '@fluentui/react-make-styles';
++import { tokens } from '@fluentui/react-theme';
+
+makeStyles({
+-  root: theme => ({ color: theme.tokenB }),
++  root: { color: tokens.tokenB },
+});
+```
+
+## Simplify types in `FluentProvider`
+
+Currently `FluentProvider` only supports the `Theme` type from `@fluentui/react-theme`. If we simplify this to `Record<string, string | number>` , we enable consumers to extend the default theme.
```

Comment:
As `FluentProvider` currently accepts only the `Theme` type, it guarantees an interoperability between a host application and a component developed in isolation - any high value component used in any host application will have access to the expected design tokens.
With changing the type to a generic record, how can we enforce the contract?

Reply:
We can optional expose `TokensProvider` (`TokensToCSSProvider`, _whatever_):

- it will do only only `useThemeStyleTag()` part
- it will have `Record<string, ...>`
- keep `FluentProvider` with `Theme`

🔽🔽🔽

```tsx
function App(props) {
  return (
    <FluentProvider theme={teamsTheme}>
      <TokensProvider tokens={customTokens}>{props.children}</TokensProvider>
    </FluentProvider>
  );
}
```

Reply:
Hmm, that might work.

Still think about the typings and the theme shape. There is one shape which you need to pass to the provider (`[cssVariableName; rawValue]` mapping) and another shape which you need to use in `makeStyles` (`[tokenName;--var(cssVariableName)]`).
Also there might be tokens which are not injected as CSS variables - those have no entries in the object you pass to the provider and have [token name; raw value] in the `makeStyles`.

Reply:
I think there is another option here where `FluentProvider` accepts a `TTheme` that extends from `Theme`. Today, `FluentProviderProps` is typed as:

```ts
export interface FluentProviderProps
  extends Omit<ComponentProps<FluentProviderSlots>, 'dir'>,
    Partial<FluentProviderCommons> {
  theme?: PartialTheme;
}
```

We could make it be:

```ts
export interface FluentProviderProps<TTheme extends PartialTheme = PartialTheme>
  extends Omit<ComponentProps<FluentProviderSlots>, 'dir'>,
    Partial<FluentProviderCommons> {
  theme?: TTheme;
}
```

In this way, we are adding extensibility to the `FluentProvider` theme shape while still maintaining that interoperability you were talking about.

Reply:
@khmakoto This is also an option. We can export a primitive component that renders variables for `Record<string, string | number>` if there will be an ask for it.

Reply:
I forgot to mention that I added snippets of this conversation to the RFC so that they are accounted for 😄